src/librustc_privacy/lib.rs

```diff
@@ -535,6 +535,324 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+                    } else if let Some(Def::Method(..)) = self.tcx.describe_def(def_id) {
+                        let candidate_impl_def_id = self.tcx.parent_def_id(def_id)
+                                                            .expect("no parent for method def_id");
+                        // `is_none` means it's an impl, not a trait
+                        if self.tcx.describe_def(candidate_impl_def_id).is_none() {
+                            impl_def_id = Some(candidate_impl_def_id)
+                        }
+                    }
+                    if let Some(impl_def_id) = impl_def_id {
+                        let self_ty = self.tcx.type_of(impl_def_id);
+                        if self_ty.visit_with(self) {
+                            return true;
+                        }
+                    }
+                }
+            }
+            ty::TyDynamic(ref predicates, ..) => {
+                let is_private = predicates.skip_binder().iter().any(|predicate| {
+                    let def_id = match *predicate {
+                        ty::ExistentialPredicate::Trait(trait_ref) => trait_ref.def_id,
+                        ty::ExistentialPredicate::Projection(proj) => proj.trait_ref.def_id,
+                        ty::ExistentialPredicate::AutoTrait(def_id) => def_id,
+                    };
+                    !self.item_is_accessible(def_id)
+                });
+                if is_private {
+                    let msg = format!("type `{}` is private", ty);
+                    self.tcx.sess.span_err(self.span, &msg);
+                    return true;
+                }
+            }
+            ty::TyAnon(def_id, ..) => {
+                for predicate in &self.tcx.predicates_of(def_id).predicates {
+                    let trait_ref = match *predicate {
+                        ty::Predicate::Trait(ref poly_trait_predicate) => {
+                            Some(poly_trait_predicate.skip_binder().trait_ref)
+                        }
+                        ty::Predicate::Projection(ref poly_projection_predicate) => {
+                            if poly_projection_predicate.skip_binder().ty.visit_with(self) {
+                                return true;
+                            }
+                            Some(poly_projection_predicate.skip_binder().projection_ty.trait_ref)
+                        }
+                        ty::Predicate::TypeOutlives(..) => None,
+                        _ => bug!("unexpected predicate: {:?}", predicate),
+                    };
+                    if let Some(trait_ref) = trait_ref {
+                        if !self.item_is_accessible(trait_ref.def_id) {
+                            let msg = format!("trait `{}` is private", trait_ref);
+                            self.tcx.sess.span_err(self.span, &msg);
+                            return true;
+                        }
+                        // Skip `Self` to avoid infinite recursion
```

Comment:
Could you `assert_eq!(trait_ref.self_ty(), ty);`? Other than that, this looks good (and we can relax it in the future).